### PR TITLE
add arg_tuple_expr

### PR DESCRIPTION
**Pull request by oxinabox**
This is useful to have and something that is needed for Nabla https://github.com/invenia/Nabla.jl/pull/189
(It is extracted out of that PR)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ Currently, this package provides the `splitdef`, `signature` and `combinedef` fu
  - `splitdef` works on a function definition expression and returns a `Dict` of its parts.
  - `combinedef` takes a `Dict` from `splitdef` and builds it into an expression.
  - `signature` works on a `Method` returning a similar `Dict` that holds the parts of the expressions that would form its signature.
-
+ - `args_tuple_expr` applies to a `Dict` from `splitdef` or `signature` to generate a expression for a tuple of it's arguments.
```
Comment (mzgubic) [inline README.md]:
```suggestion
 - `args_tuple_expr` applies to a `Dict` from `splitdef` or `signature` to generate an expression for a tuple of its arguments.
```

File: README.md [hunk 2]
```diff
@@ -38,14 +38,18 @@ Dict{Symbol,Any} with 5 entries:
...
+
 julia> def[:name] = :g;
 
 julia> def[:head] = :(=);
 
-julia> def[:body] = :(x * y);
+julia> args_tuple_expr(def)
+:((x, y))
+
+julia> def[:body] = :(*($(args_tuple_expr(def))...));
```
Comment (mzgubic) [inline README.md:52]: is this right?

File: src/def_tools.jl [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# These utilities are for working with the signature_def `Dict` that comes out of
+# `signature`/`splitdef`
+
+
+
+"""
+    args_tuple_expr(signature_def::Dict{Symbol})
+    args_tuple_expr(arg_exprs)
+
+For `arg_exprs` being a list of arguments expressions from a signature, of a form
```
Comment (mzgubic) [inline src/def_tools.jl]:
```suggestion
For `arg_exprs` being a list of argument expressions from a signature, of a form
```

File: src/def_tools.jl [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+such as `[:(x::Int), :(y::Float64), :(z::Vararg)]`, or being a whole signature_def `Dict`
+containing a `signature_def[:args]` value of that form.
+
+This returns a tuple expresion containing all of the args by name; while correctly handling
+splatting, for things that are `Vararg` typed. e.g for the prior example `:((x, y, z...))`
```
Comment (mzgubic) [inline src/def_tools.jl]:
```suggestion
This returns a tuple expresion containing all of the args by name. It correctly handles
splatting for things that are `Vararg` typed, e.g for the prior example `:((x, y, z...))`
```

File: src/def_tools.jl [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+```julia
+signature_def[:body] = quote
+    args = \$(args_tuple_expr(signature_def))
+    println("args = ",args)
+    \$(signature_def[:body]) # insert old body
+end
+```
+
+A more realistic use case is if you want to insert a call to a another function
+that accepts same argument as the original function.
```
Comment (mzgubic) [inline src/def_tools.jl]:
```suggestion
A more realistic use case is if you want to insert a call to another function
that accepts the same arguments as the original function.
```

File: src/def_tools.jl [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    ret.args = map(arg_exprs) do arg
+        arg isa Symbol && return arg  # it is just `x` not `x::T`
+
+        # remove splatting (will put back on at end)
+        was_splatted = Meta.isexpr(arg, :(...), 1)
+        if was_splatted
+            arg = arg.args[1]
+        end
+
+        # handle presence or absense of type constrantain
```
Comment (mzgubic) [inline src/def_tools.jl]:
```suggestion
        # handle presence or absence of type constraints
```

File: src/def_tools.jl [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+that accepts same argument as the original function.
+"""
+function args_tuple_expr end
+
+args_tuple_expr(signature_def::Dict{Symbol}) = args_tuple_expr(signature_def[:args])
+
+function args_tuple_expr(arg_exprs)
+    ret = Expr(:tuple)
+    ret.args = map(arg_exprs) do arg
+        arg isa Symbol && return arg  # it is just `x` not `x::T`
```
Comment (mzgubic) [inline src/def_tools.jl]: can we just delete this since it is handled below? or do we want to short circuit here?

File: src/def_tools.jl [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            arg_name, Texpr = arg.args
+        elseif arg isa Symbol
+            arg_name = arg
+            Texpr = nothing
+        else
+            error("unexpected form of argument: $arg")
+        end
+
+        # Clean up types so we can recognise if it is `Vararg`
+        # remove where clauses (because the interfere with recongnizing Vararg)
```
Comment (mzgubic) [inline src/def_tools.jl]:
```suggestion
        # remove where clauses (because they interfere with recognizing Vararg)
```

**Review by mzgubic — APPROVED** [review]
i was going to complain it doesn't deal with kwargs, but that seems by design (and maybe addressed by other PRs, haven't yet checked)

only one real comment, the rest is just fixing typos

File: src/def_tools.jl (hunk 7 above)
Comment (oxinabox) [reply src/def_tools.jl]: Oh, yeah, i forgot to delete this, yes we can.
I added features incrementally and minimally, one test at a time.

```suggestion
```

**Comment by oxinabox** [comment]
> i was going to complain it doesn't deal with kwargs, but that seems by design

This is by design, kwargs are not a feature I need right now.
I will add to the docstring that it only does positional arguments

File: README.md (hunk 2 above)
Comment (oxinabox) [reply README.md:52]: why wouldn't it be.
The next line looks right to me.

File: src/def_tools.jl (hunk 3 above)
Comment (oxinabox) [reply src/def_tools.jl]:
```suggestion
For `arg_exprs` being a list of positional argument expressions from a signature, of a form
```

**Pull request merged**